libs/pika/async_mpi/include/pika/async_mpi/dispatch_mpi.hpp

```diff
@@ -0,0 +1,298 @@
+                    pika::detail::try_catch_exception_ptr(
+                        [&]() mutable {
+                            using namespace pika::debug::detail;
+                            using ts_element_type = std::tuple<std::decay_t<Ts>...>;
+                            using invoke_result_type = mpi_request_invoke_result_t<F, Ts...>;
+                            using namespace pika::debug::detail;
+                            PIKA_DETAIL_DP(mpi_tran<5>,
+                                debug(str<>("dispatch_mpi_recv"), "set_value_t", "stream",
+                                    detail::stream_name(r.op_state.stream_)));
+
+                            // move/copy the received params into our local opstate
+                            r.op_state.ts.template emplace<ts_element_type>(
+                                PIKA_FORWARD(Ts, ts)...);
+                            // and get a reference to the tuple of local params
+                            auto& t = std::get<ts_element_type>(r.op_state.ts);
+                            // init a request
+                            MPI_Request request;
+                            int status = MPI_SUCCESS;
+                            // invoke the function, with the contents of the param tuple as args
+                            pud::invoke_fused(
+                                [&](auto&... ts) mutable {
+                                    // execute the mpi function call, passing in the request object
+                                    if constexpr (std::is_void_v<invoke_result_type>)
+                                    {
+                                        PIKA_INVOKE(PIKA_MOVE(r.op_state.f), ts..., &request);
+                                    }
+                                    else
+                                    {
+                                        status =
+                                            PIKA_INVOKE(PIKA_MOVE(r.op_state.f), ts..., &request);
+                                    }
+                                    PIKA_DETAIL_DP(mpi_tran<7>,
+                                        debug(str<>("dispatch_mpi_recv"), "invoke mpi",
+                                            detail::stream_name(r.op_state.stream_), ptr(request)));
+
+                                    PIKA_ASSERT_MSG(request != MPI_REQUEST_NULL,
+                                        "MPI_REQUEST_NULL returned from mpi invocation");
+                                },
+                                t);
+                            if (status != MPI_SUCCESS)
+                            {
+                                PIKA_DETAIL_DP(mpi_tran<5>,
+                                    debug(str<>("set_error"), "status != MPI_SUCCESS",
+                                        detail::error_message(status)));
+                                exp::set_error(PIKA_MOVE(r.op_state.receiver),
+                                    std::make_exception_ptr(mpi_exception(status)));
```

Comment:
since there is already a check for `is_void`, it is enough to put the static assert `<int>` in the `else` part.
For now I'd prefer to leave the exceptions, they have proven to be useful so far finding bugs/errors as the mpi error message is displayed and if we only return an int status, then the user has to do the checking. I'll happily redo this for the next patch if necessary.